### PR TITLE
core(tracehouse): use tasks to improve profiler timing data

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
The next phase of CPU profiler work got really out of hand really quickly, so stopping at a reasonable checkpoint. This PR improves the timing of sampling profiler data using other task information from a trace. The sampling profiler is still out of the normal flow of Lighthouse, so the test cases are the only actual usage.

Also a nice juicy debug function for task data which should help generate those nice "artistic" rendering of traces :)

**Related Issues/PRs**
ref #8526

### CODE REVIEW

File: lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js [hunk 1]
```diff
@@ -65,6 +65,40 @@ describe('CPU Profiler Model', () => {
...
+        {ph: 'B', ts: ts(80e2), args: {data: {callFrame: {functionName: 'Foo'}}}},
+        {ph: 'B', ts: ts(120e2), args: {data: {callFrame: {functionName: 'Bar'}}}},
+        {ph: 'B', ts: ts(125e2), args: {data: {callFrame: {functionName: 'Baz'}}}},
+        {ph: 'E', ts: ts(134e2), args: {data: {callFrame: {functionName: 'Baz'}}}},
+        {ph: 'E', ts: ts(157e2), args: {data: {callFrame: {functionName: 'Bar'}}}},
+        {ph: 'B', ts: ts(168e2), args: {data: {callFrame: {functionName: 'Baz'}}}},
+        {ph: 'E', ts: ts(175e2), args: {data: {callFrame: {functionName: 'Baz'}}}},
+        {ph: 'E', ts: ts(175e2), args: {data: {callFrame: {functionName: 'Foo'}}}},
+        {ph: 'E', ts: ts(190e2), args: {data: {callFrame: {functionName: '(program)'}}}},
+        {ph: 'E', ts: ts(190e2), args: {data: {callFrame: {functionName: '(root)'}}}},
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js:98]: that's what the comments above every line in `tasks` var were supposed to do. I'm not sure what I would do to make this better :/ Are you asking for narrower assertions that only look at the differences between this and the "events in order" test?

File: lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js [hunk 2]
```diff
@@ -65,6 +65,40 @@ describe('CPU Profiler Model', () => {
       ]);
     });
 
+    it('should create events while aware of other tasks', () => {
+      const ts = x => 9e6 + x;
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js:69]: yep!

File: lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js [hunk 3]
```diff
@@ -65,6 +65,40 @@ describe('CPU Profiler Model', () => {
       ]);
     });
 
+    it('should create events while aware of other tasks', () => {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js:68]: sweet thanks, when we get to the skipped sample logic we'll definitely incorporate those 👍

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 4]
```diff
@@ -176,16 +278,38 @@ class CpuProfilerModel {
...
+      } = this._findEffectiveTimestamp({
+        earliestPossibleTimestamp: lastEffectiveTimestamp,
+        latestPossibleTimestamp: profilerTimestamp + timeDelta,
+        lastKnownTaskStartTimeIndex,
+        lastKnownTaskEndTimeIndex,
+        knownTasksByStartTime,
+        knownTasksByEndTime,
+      });
+
+      profilerTimestamp += timeDelta;
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:295]: this is just a rename of the old `timestamp` which is the raw timestamp of a sample as reported by the profiler's timeDeltas, but it sounds like the new name made that less clear to you :)

`currentTimestampAsReportedByProfiler`
`rawTimestamp`
`currentRawTimestamp`

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 5]
```diff
@@ -155,19 +155,121 @@ class CpuProfilerModel {
...
+   *
+   * With the sampling profiler we know that a function started/ended *sometime between* two points,
+   * but not exactly when. Using the information from other tasks gives us more information to be
+   * more precise with timings and allows us to create a valid task tree later on.
+   *
+   * @param {Array<LH.Artifacts.TaskNode>} [knownTasks]
    * @return {Array<LH.TraceEvent>}
    */
-  createStartEndEvents() {
+  createStartEndEvents(knownTasks = []) {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:258]: sure sg 👍

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 6]
```diff
@@ -155,19 +155,121 @@ class CpuProfilerModel {
...
+   *
+   * Because the sampling profiler only provides a *range* of start/stop function boundaries, this
+   * method uses knowledge of a known set of tasks to find the most accurate timestamp for a particular
+   * range. For example, if we know that a function ended between 800ms and 810ms, we can use the
+   * knowledge that a toplevel task ended at 807ms to use 807ms as the correct endtime for this function.
+   *
+   * @param {{earliestPossibleTimestamp: number, latestPossibleTimestamp: number, lastKnownTaskStartTimeIndex: number, lastKnownTaskEndTimeIndex: number, knownTasksByStartTime: Array<{startTime: number, endTime: number}>, knownTasksByEndTime: Array<{startTime: number, endTime: number}>}} data
+   * @return {{timestamp: number, lastStartTimeIndex: number, lastEndTimeIndex: number}}
+   */
+  _findEffectiveTimestamp(data) {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:199]: > doing it higher up at either the cpu profiler model level or the main thread tasks level

that's exactly what the unit test added in this PR does ;)

I can add explicit unit tests for this private method as well if you think the current cases aren't sufficient 👍

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 7]
```diff
@@ -155,19 +155,121 @@ class CpuProfilerModel {
...
   /**
-   * Creates B/E-style trace events from a CpuProfile object created by `collectProfileEvents()`
+   * Finds all the tasks that started or ended (depending on `type`) within the provided time range.
+   * Uses a memory index to remember the place in the array the last invocation left off to avoid
+   * re-traversing the entire array.
    *
+   * @param {Array<{startTime: number, endTime: number}>} knownTasks
+   * @param {{type: 'startTime'|'endTime', lastKnownTaskIndex: number, earliestPossibleTimestamp: number, latestPossibleTimestamp: number}} options
+   */
+  _getTasksInRange(knownTasks, options) {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:165]: would you like all of them moved or just the new ones? I'm a little concerned about muddying the diff with a move of all private methods here

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 8]
```diff
@@ -598,6 +598,83 @@ class MainThreadTasks {
...
+    const tasksByDepth = new Map();
+    for (const task of tasks) {
+      if (task.startTime > endTime || task.endTime < startTime) continue;
+
+      const depth = computeTaskDepth(task);
+      const tasksAtDepth = tasksByDepth.get(depth) || [];
+      tasksAtDepth.push(task);
+      tasksByDepth.set(depth, tasksAtDepth);
+
+      const id = String.fromCharCode(65 + (taskLegend.size % 26));
```
Comment (patrickhulce) [inline lighthouse-core/lib/tracehouse/main-thread-tasks.js:641]:
```suggestion
      // Create a user-friendly ID for new tasks using a capital letter.
      // 65 is the ASCII code for 'A' and there are 26 letters in the english alphabet.
      const id = String.fromCharCode(65 + (taskLegend.size % 26));
```

File: lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js [hunk 9]
```diff
@@ -805,14 +805,41 @@ describe('Main Thread Tasks', () => {
...
+        {ph: 'X', name: 'TaskC', pid, tid, ts: baseTs + 10e3, dur: 30e3, args},
+        {ph: 'E', name: 'TaskB', pid, tid, ts: baseTs + 55e3, args},
+        {ph: 'X', name: 'TaskD', pid, tid, ts: baseTs + 60e3, dur: 30e3, args},
       ];
 
       traceEvents.forEach(evt => Object.assign(evt, {cat: 'devtools.timeline'}));
 
-      expect(() => run({traceEvents})).toThrow();
+      const tasks = run({traceEvents});
+      const taskTreeAsString = MainThreadTasks.printTaskTreeToDebugString(tasks, {printWidth: 50});
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/test/lib/tracehouse/main-thread-tasks-test.js:828]: > it doesnt get any use in this PR aside from its own test. what's the plan?

Use it like `console.log` :) (like I did when visualizing all of the profiler tasks when making this PR). I can't think of a another use for it that wouldn't be commented out in code on master and isn't test-specific.

> do you think it oversteps to create main thread tasks from the new cpu profiler trace events to visualize that test case? it'd obv be making main thread tasks when we dont necc need it, but having the visual is a nice way to visually assert the result.

I like it 👍

File: lighthouse-core/lib/tracehouse/main-thread-tasks.js [hunk 10]
```diff
@@ -598,6 +598,83 @@ class MainThreadTasks {
...
   }
+
+  /**
+   * Prints an artistic rendering of the task tree for easier debugability.
+   *
+   * @param {TaskNode[]} tasks
+   * @param {{printWidth: number, startTime?: number, endTime?: number, taskLabelFn?: (node: TaskNode) => string}} options
+   * @return {string}
+   */
+  static printTaskTreeToDebugString(tasks, options) {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/main-thread-tasks.js:609]: My initial impression is that it doesn't seem worth it to move away from the place that the debugging helps, but a followup for relocating all of our debugging utils sgtm

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js]: went with currentProfilerTimestamp

**Review by paulirish — APPROVED** [review]

**Pull request merged**